### PR TITLE
[Perf] Batch Weight Prefetching via cuMemcpyBatchAsync to Reduce Latency

### PR DESCRIPTION
**Pull request by xiaobao520123**
## Purpose

Current implementation of weight prefetching requires launching `cuMemcpyAsync` multiple times, which creates driver-call overhead. This overhead accumulates and hurts inference performance because prefetching is hooked up near the end of each layer `forward`.

Replace per-layer weight prefetching with `swap_blocks_batch`. `swap_blocks_batch` utilizes `cuMemcpyBatchAsync` since CUDA 12.8+. According to the implementation, `swap_blocks_batch` will fallback to `cuMemcpyAsync` loop when the feature is not available.

In addition, objects related to CUDA stream synchronization have been optimized to further reduce the overhead.

As a result, it yields **>50%** latency decrease in triggering prefetching and reduces both TTFT and TPOT/ITL (See the test result below).

### Original Trace
Origin_CUDA_Graph_Trace

### Trace after This PR
PR_CUDA_Graph_Trace

### Hardware
**GPU**: 1xA100-40GB 40GB, CUDA 12.8/12.9
**CPU**: 2× Intel Xeon Gold 5318Y @ 2.10 GHz (24 cores/socket, 2 threads/core → 96 logical CPUs, 48 physical cores), max turbo 3.4 GHz, 2 NUMA nodes.
**GPU ↔ CPU**: PCIe Gen4 x16 (30GB/s R/W theoretical, 25GB/s actual)
**Memory**:  ~690 GiB usable (724 GB raw); DDR4-3200 ECC, running at 2933 MT/s

### vLLM
**Version**:
- `vllm==0.19.1rc1.dev750+ga3ec4a35f.precompiled`
- `VLLM_USE_PRECOMPILED=1 VLLM_PRECOMPILED_WHEEL_COMMIT=3527229517f01a5f2406fa6fbf35ff9223c65ed5 uv pip install --editable . --torch
-backend=cu129`.
- Last Git Commit: `a3ec4a35f5943c250974d504706d22297d423468 [Bugfix][Metrics] Fix RayPrometheusMetric.labels() returning shared labeled child (#40840)`

**Configuration**:
- Model: Qwen3-8B/Qwen3-32B
- Max model length: 24k
- TP: 1
- GPU util: 90%
- Offload backend: `prefetch`
- Offload groups:
  - Qwen3-8B (Dense, 36 layers): `50%` offload; 18 groups, 2 layers each, 1 layer offloaded, ratio: `2:1`; Prefetch steps: 2
  - Qwen3-32B (Dense, 64 layers): `75%` offload; 16 groups, 4 layers each, 3 layers offloaded, ratio: `4:3`; Prefetch steps: 2

vLLM Instance:
```
vllm serve Qwen/Qwen3-8B \
      --port 8000 \
      --tensor-parallel-size 1 \
      --gpu-memory-utilization 0.9 \
      --max-model-len 24000 \
      --offload-backend prefetch \
      --offload-group-size 2 \
      --offload-num-in-group 1 \
      --offload-prefetch-step 2
```

**Dataset & Benchmark**:
Dataset: LongBench-NarrativeQA (Shuffled): [Link](https://huggingface.co/datasets/zai-org/LongBench#:~:text=200-,NarrativeQA,-Single%2Ddoc%20QA)
QPS: 1
Prompts: 10

Benchmark & Torch Profiling:
```
vllm bench serve \
  --backend vllm \
  --model Qwen/Qwen3-8B \
  --dataset-name custom \
  --dataset-path /workspace/projects/datasets/narrativeqa_shuffled.jsonl \
  --skip-chat-template \
  --disable-shuffle \
  --num-prompts 10 \
  --request-rate 1 \
  --profile \
  --result-dir [PLACEHOLDER] \
  --result-filename [PLACEHOLDER] \
  --save-result
```

#### (Major Improvement) Per-prefetch-call Host-side Latency
Mode | Variant | Calls | Mean (μs) | Median (μs) | P99 (μs) | Total host (ms)
-- | -- | -- | -- | -- | -- | --
CUDA Graph Enabled | Original | 864 | 353.4 | 348.2 | 432.8 | 305.3
CUDA Graph Enabled | PR | 864 | 159.4 | 151.9 | 181.5 | 137.7
  |   |   | -54.9% | -56.4% | -58.1% | -54.9%
Eager | Original | 5,472 | 342.8 | 334.4 | 478.9 | 1876.1
Eager | PR | 5,472 | 149.8 | 148.4 | 177.3 | 819.6
  |   |   | -56.3% | -55.6% | -63.0% | -56.3%
##### CUDA Graph Enabled
Torch Profile (Original): [Link](https://huggingface.co/datasets/xiaobao520123/pasa-exp/blob/main/torch/profile/weight_prefetch_qwen3-8B_longbench_all/origin/cuda_graph/weight_prefetch_shuffled_1qps/rank0.1777648073798165136.pt.trace.json.gz)
Torch Profile (cuMemcpyBatchAsync): [Link](https://huggingface.co/datasets/xiaobao520123/pasa-exp/blob/main/torch/profile/weight_prefetch_qwen3-8B_longbench_all/fixed/cuda_graph/weight_prefetch_shuffled_1qps/rank0.1777632985348306463.pt.trace.json.gz)
Metric | Unit | Original | This PR | Δ
-- | -- | -- | -- | --
Output tokens | — | 2 560 | 2 560 | 0
Request throughput | req/s | 0.1161 | 0.1161 | -0.03%
Output throughput | tok/s | 29.72 | 29.71 | -0.03%
Total token throughput | tok/s | 1567.5 | 1567.0 | -0.03%
Duration | s | 86.14 | 86.17 | +0.03%
Mean TTFT | ms | 4927.3 | 4912.2 | -0.31%
Median TTFT | ms | 5405.0 | 5380.3 | -0.46%
P99 TTFT | ms | 8412.0 | 8392.1 | -0.24%
Mean TPOT | ms | 279.59 | 279.74 | +0.05%
Mean ITL | ms | 279.59 | 279.74 | +0.05%
##### Eager
Torch Profile (Original): [Link](https://huggingface.co/datasets/xiaobao520123/pasa-exp/blob/main/torch/profile/weight_prefetch_qwen3-8B_longbench_all/origin/eager/weight_prefetch_shuffled_1qps/rank0.1777648371496668212.pt.trace.json.gz)
Torch Profile (cuMemcpyBatchAsync): [Link](https://huggingface.co/datasets/xiaobao520123/pasa-exp/blob/main/torch/profile/weight_prefetch_qwen3-8B_longbench_all/fixed/eager/weight_prefetch_shuffled_1qps/rank0.1777633270305561051.pt.trace.json.gz)
Metric | Unit | Original | This PR | Δ
-- | -- | -- | -- | --
Output tokens | — | 2 560 | 2 560 | 0
Request throughput | req/s | 0.1154 | 0.1161 | +0.59%
Output throughput | tok/s | 29.54 | 29.72 | +0.59%
Total token throughput | tok/s | 1558.2 | 1567.4 | +0.59%
Duration | s | 86.65 | 86.14 | -0.59%
Mean TTFT | ms | 5712.3 | 5446.2 | -4.66%
Median TTFT | ms | 6185.1 | 5903.3 | -4.56%
P99 TTFT | ms | 9214.3 | 8938.7 | -2.99%
Mean TPOT | ms | 278.63 | 277.75 | -0.32%
Mean ITL | ms | 278.63 | 277.75 | -0.32%

### Qwen3-32B (Profiling disabled)
* Qwen3-32B is too large to fit the hardware, and weight offloading/prefetching pushes significant pressure on memory & bandwidth. Therefore, performace gain is realtively small.
##### CUDA Graph Enabled
Metric | Unit | Original | This PR | Δ
-- | -- | -- | -- | --
Output tokens | — | 2 498 | 2 150 | **-14%**
Request throughput (normalized) | req/s | 0.01022 | 0.01022 | +0.005%
Output throughput (normalized) | tok/s | 2.5520 | 2.5521 | +0.005%
Total token throughput (normalized) | tok/s | 137.88 | 137.89 | +0.005%
Duration (normalized to 2 498 tok) | s | 978.86 | 978.81 | -0.005%
Mean TTFT | ms | 251 829 | 234 892 | -6.7%
Median TTFT | ms | 244 395 | 98 586 | -60%
P99 TTFT | ms | 498 065 | 527 317 | +5.9%
Mean TPOT | ms | 1 859.03 | 1 858.99 | -0.002%
Mean ITL | ms | 1 859.05 | 1 858.73 | -0.017%

##### Eager
Metric | Unit | Original | This PR | Δ
-- | -- | -- | -- | --
Output tokens | — | 2 192 | 2 192 | 0
Request throughput | req/s | 0.009942 | 0.009944 | +0.02%
Output throughput | tok/s | 2.1794 | 2.1798 | +0.02%
Total token throughput | tok/s | 133.880 | 133.905 | +0.02%
Duration | s | 1005.80 | 1005.61 | -0.02%
Mean TTFT | ms | 228 152.3 | 228 059.9 | -0.04%
Median TTFT | ms | 57 661.4 | 57 594.2 | -0.12%
P99 TTFT | ms | 526 514.9 | 526 376.7 | -0.03%
Mean TPOT | ms | 1863.00 | 1862.74 | -0.014%
Mean ITL | ms | 1863.27 | 1863.01 | -0.014%

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/offloader/prefetch.py [hunk 1]
```diff
@@ -390,6 +391,10 @@ def __init__(
         # Used for per-layer synchronization (both eager and capture modes).
         self._copy_done_event = torch.cuda.Event()
 
+        # Fork: record event on compute stream, copy_stream waits on it.
+        # This joins copy_stream to any active CUDA graph capture.
+        self._fork_event = torch.cuda.Event()
```
[the remark this answers is not in the document]
Comment (xiaobao520123) [reply vllm/model_executor/offloader/prefetch.py:396]: This should align with the original implementation. Won't explictly set this parameter.

**Comment by minosfuture** [comment]
Thanks for the work and detailed data!
Prefetching should be completely hidden behind computation. This would be the case if batch size is large (as it is in prefill). It may be hard on A100 nodes that doesn't have high-bandwidth C2C.
Is this mostly addressing the inter-forward-step prefetching for the first offloaded-layer? What's percentage of the delay on critical path over the whole step latency and why is it accumulating?
CUDA graph incompatibility of cuMemcpyBatchAsync is not great.  But it should be compatible with torch.compile?

**Comment by xiaobao520123** [comment]
@minosfuture, thanks for the question!

> Prefetching should be completely hidden behind computation.

Yes, it should. When doing data transfer, current implementation runs in async. Yet, such prefetching on future layers starts at the end after computation on previous layer is complete.
```Python
# prefetch.py:210-242
def _hook_module_forward(self, index: int, module: nn.Module):
    original_forward = module.forward

    def forward(*args, **kwargs):
        module.forward = original_forward
        torch.ops.vllm.wait_prefetch(input_tensor, index)

        output = original_forward(*args, **kwargs)

        next_index = (index + self.prefetch_step) % len(self.module_offloaders)

        # NOTICE: start prefetch task immediately after forward completes. 
        if isinstance(output, tuple):
            torch.ops.vllm.start_prefetch(output[0], next_index)
        else:
            torch.ops.vllm.start_prefetch(output, next_index)

        module.forward = forward
        return output

    module.forward = forward
```
The startup is hooked onto the inference critical path (forward) and because it takes API calls for the CPU to submit tasks to the GPU. This leads to overhead. Keeping this 'hook' scheme, we must reduce such overhead and avoid run bubbles as much as possible.

> This would be the case if batch size is large (as it is in prefill). It may be hard on A100 nodes that doesn't have high-bandwidth C2C.

Agree. That’s why we can see bigger latency decrease on TTFT, even bigger in eager mode. One of the biggest application of offloading is when only commodity hardware is available, or running in single node instance. The work is trying to reduce any other factors regardless it is compute/memory/bandwidth-bound.

> Is this mostly addressing the inter-forward-step prefetching for the first offloaded-layer?

The savings are **per-prefetch-call** of inter-forward-steps, applied to every start_prefetch invocation on the critical path, not just the first layer in a step.

> What's percentage of the delay on critical path over the whole step latency and why is it accumulating?

In my benchmark, depending on offload_prefetch_step settings, one prefetch launch call stands `~35%` of the latency on each layer forward step, and it adds up total `~2.2%` of delays on one model execution. `start_prefetch` is invoked synchronously inside the layer's Python forward hook, between the previous layer's compute and the next kernel launch. Every microsecond spent on the host here directly delays issuing the next kernel — it cannot be hidden behind GPU compute, even when the H2D copy `(cuMemcpyAsync)` itself can.

> CUDA graph incompatibility of cuMemcpyBatchAsync is not great. But it should be compatible with torch.compile?

I hope CUDA official will support it in the future. Nonetheless, everything is still compatible with `torch.compile()`.

**Comment by xiaobao520123** [comment]
Reference:
1. #29941
2. #38460
